Assignment 1 PR for Feedback

gicamikiel opened the pull request:
UPDATE: Reimplementation completed, commits linked below
https://github.com/gicamikiel/Computing-at-Scale/commit/d7c9b3837a2a8f3f9e91bb7566c9f663f526abec
https://github.com/gicamikiel/Computing-at-Scale/commit/8250d05820a0bd4ba4ce5967b921278be0675651
(no one brought this up but this was bugging me)
https://github.com/gicamikiel/Computing-at-Scale/commit/c10b254b8d661e2d341475374498473fd099374b

gicamikiel committed:
Initial commit

gicamikiel committed:
* PolynomialFunction implementation passing tests
* Files renamed

gicamikiel committed:
Integration rule classes

gicamikiel committed:
Quadrature integration tests

gicamikiel committed:
Numerical integration unit tests

gicamikiel committed:
polynomial_integrate implemented

gicamikiel committed:
Create README.md

gicamikiel committed:
Create assignment1.md

gicamikiel committed:
Update README.md

gicamikiel committed:
Revert "Update README.md"

This reverts commit 2fc2e7fdf48c0fd6d8faa05fedc2f1189600713a.

ickbumk commented on assignment1/README.md:
Everything compiled nicely, and the instructions were clear. The unittest works nicely. Chebyshev fails for all cases.

ickbumk commented on assignment1/source/integrationrule.cpp:
```diff
@@ -0,0 +1,70 @@
+
+double GaussLobatto::operator()(ScalarFunction& function, double from, double to) const {
+    double out = 0;
+    for (int i=1; i<num_pts()-1; i++) { // min number of num_pts is 3
```
std::size_t (I think) would better suit for indexing

scottblender replied:
you could also use auto.

ickbumk commented on assignment1/source/integrationrule.cpp:
```diff
@@ -0,0 +1,70 @@
+
+double IntegrationRule::operator()(ScalarFunction& function, double from, double to) const {return 0;}
+
+GaussLegendre::GaussLegendre() : xi_pts_({}), weights_({}) {}
```
I think the constructor does not need empty {} since the default is an empty state?

gicamikiel replied:
It seems like a get a compile error if I leave out the empty {}

ickbumk commented on assignment1/source/integrationrules.hpp:
```diff
@@ -0,0 +1,30 @@
+
+// File for defining common quadrature rules
+
+#define HI_WT_PT std::sqrt(3.0/7 - 2.0/7*std::sqrt(6.0/5))
```
we can try "constexpr" instead of #define

gicamikiel replied:
I used macros here because I did not want these to be used outside of the file.

Macros are fairly ugly though and I realized I might be able to limit this variable to one file using an anonymous namespace? I will try a constexpr variable in one.

ickbumk commented on assignment1/source/polynomial_integrate.cpp:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    double out = (*rule)(p, from, to);
+    std::cout << out << "\n";
```
return 0 maybe?

ickbumk reviewed (commented):
Does the style match?
Yes

Is there documentation of the new code? Does it address what it does and how to use it?

I think the code will definitely improve with more module docstring and function docstrings. I was able to follow the code without significant difficulties, just because I did the same assignment. However, if someone who is not familiar with the assignment may have hard time understanding what's going on.

Are there any obvious bugs, or performance issues?

No bugs or performance issues.

Is the code easy to read and understand? Is it well organized? Are function names clear?

all function names were clear. No abbreviations were used. The header files are well organized .

Can raw loops be replaced with standard library algorithms?

Nothing that I am aware of

General questions or concerns.

No concerns. I am impressed how the unittest was constructed. My implementation is far primitive than this. 


Does the code compile?

Yes

Are there any compiler warnings?

No

Are there any runtime errors?

No

Does the code do what it is supposed to do?

Yes. The unittest also performs test for all possible cases. 

Are there test cases for all new functionality?

Unittest thoroughly tests for all cases.

Are there test cases that you think should be added?

No

Do all of the tests pass?

No. Chebyshev does not pass the test. 

Are there any performance regressions?

Not that I am aware of

Does the code scale to the expected problem size?

We have not tried this problem with more than 1 dimensional case or polynomials with very high degree. This is not applicable.

Great code overall. Impressive.

scottblender reviewed (commented):
- Does the style match the rest of the codebase? (naming conventions, indentation, etc.) Tools like clang-format can help.
    - Yes, the style is consistent.
- Is there documentation of the new code? Does it address what it does and how to use it?
    - Yes, the documentation is fine. The README.md could include much more information, though, about how each base class and derived class works, how to input information into how to use polynomial_integrate, and what methods/functions are called.
    - Code can also have more comments to explain functionality, even if it should be easy to read/understand.
- Are there any obvious bugs or performance issues?
    - No, there are no apparent bugs or performance issues.
- Is the code easy to read and understand? Is it well-organized? Are function names clear?
    - Yes, the code is easy to read and well-organized.
- Can raw loops be replaced with standard library algorithms?
    - Mostly no. It might be worthwile to look at `std::for_each` on cppreference.
- Does the code compile?
    - Yes.
- Are there any compiler warnings?
    - No.
- Are there any runtime errors?
    - No.
- Does the code do what it is supposed to do?
    - Yes.
- Are there test cases for all new functionality?
    - Yes.
- Are there test cases that you think should be added? i.e., corner cases that are not covered.
    - Gauss-Legendre and Gauss-Lobatto work well. For Gauss-Chebyshev, since the weighting function is sqrt(1-x^2) or 1/sqrt(1-x^2), I would modify this to only work to integrate functions of this type in the specified bounds. I believe the change of coordinates for this quadrature method is a bit more complex since h(x) = f(x)/sqrt(1-x^2) and dh/dxi becomes slightly more complicated when scaling the integral. 
    - For the specific test cases, I like that you tested all polynomials the quadrature method can exactly integrate.  In your specific quadrature methods, one thing to think about is throwing a `std::invalid_argument` if the incorrect number of points are sent in or if the polynomial is out of bounds.
- Do all of the tests pass?
    - No. Chebyshev is incorrect, but this can be solved by simplfying it to only work for -1 to 1.
- Are there any performance regressions?
    - No, performance is fine. To improve performance, it might be better to test edge cases rather than every single possible polynomial. 
- Does the code scale to the expected problem size?
    - Yes, the code is designed to scale well.

scottblender commented on assignment1/README.md:
````diff
@@ -0,0 +1,9 @@
+3. Run ```cmake --build build``` to compile
+
+## Usage
+* Run ```build/source/polynomial_integrate``` to get usage information for integrating polynomials in the command line.
````
What are the inputs and input orders for this? Could you provide instructions in the README.md in case someone trying to use this could not see the source code?

gicamikiel replied:
I relied on writing and printing a good help message when the integrate polynomial executable was run, which means you technically do not need to look at the source code. 

When people use new software on Linux they usually check the help message when you run the executable, but you are right that I should add instructions in the README as it would be better.

scottblender commented on assignment1/source/integrationrule.cpp:
```diff
@@ -0,0 +1,70 @@
+}
+
+GaussChebyshev::GaussChebyshev(int num_pts) {
+    if(num_pts <= 0) throw std::invalid_argument("Error: cannot have negative or zero number of points");
```
I would add in cases where you enter more than 4 points too.,

scottblender commented on assignment1/source/integrationrules.hpp:
I would rename this to something a little more clear. It might be confusing for someone looking at both the base class and the derived class.

scottblender commented on assignment1/source/integrationrules.hpp:
```diff
@@ -0,0 +1,30 @@
+#define LOW_WT_PT std::sqrt(3.0/7 + 2.0/7*std::sqrt(6.0/5))
+#define LOW_WT (18-std::sqrt(30))/36.0
+
+std::unordered_map<int, GaussLegendre> gl_rules = {
```
nice use of `std::unordered_map`

scottblender commented on assignment1/source/scalarfunction.cpp:
```diff
@@ -0,0 +1,35 @@
+}
+
+double& PolynomialFunction::operator[](int exp)  {
+    if (exp < 0) throw std::invalid_argument("Error: polynomials cannot have negative exponents");
```
Nice check for negative exponents. What about fractional exponents, too?

gicamikiel replied:
That is a good idea, I did not think about fractional exponents

gicamikiel replied:
Actually I just realized passing in a fractional exponent should not be possible because the parameter data type is an integer

scottblender commented on assignment1/source/integrationrule.cpp:
Using templates for the return type of the () operators or wherever you initialize a std::vector might be worthwhile.

scottblender commented on assignment1/source/integrationrule.hpp:
```diff
@@ -0,0 +1,56 @@
+        int num_pts() const;
+
+    protected: 
+        std::vector<double> const xi_pts_;
```
Should these be const? Might the user want to change the values of these after initializing them?

gicamikiel replied:
The intent I was going for here was that if someone needs the values changes they can create another instance with different values. I don't expect lots of IntegrationRule classes to be created on the fly so this should not be an issue. Maybe I could have made this clearer.

scottblender commented on assignment1/source/integrationrules.hpp:
```diff
@@ -0,0 +1,30 @@
+    {4, GaussLegendre({-LOW_WT_PT, -HI_WT_PT, HI_WT_PT, LOW_WT_PT}, {LOW_WT, HI_WT, HI_WT, LOW_WT})}
+};
+
+std::map<int, GaussLobatto> glo_rules = {
```
Why use a regular `std::map` compared to a `std::undordered_map` for Gauss-Lobatto?

gicamikiel replied:
They should be the same, I think I made a mistake here

scottblender commented on assignment1/source/polynomial_integrate.cpp:
```diff
@@ -0,0 +1,65 @@
+    GaussChebyshev gc(num_pts);
+    if(rulename == "gauss-legendre") {
+        if(num_pts > 4) {
+            std::cout << "Only 1-4 points are supported\n";
```
I think implementing this in the class itself is "safer."

gicamikiel replied:
I kept the definition for the maximum number of points outside of any classes because a user of the library can technically define more integration points for their own purposes.

gicamikiel committed:
Merge branch 'main' into assignment1

gicamikiel committed:
Improve documentation for polynomial_integrate

gicamikiel committed:
Minor refactors and style updates

gicamikiel committed:
Factor out the entire Gauss-Lobatto rule class as providing the appropriate
weights and points are sufficient to define it.

Rename GaussLegendre class to GaussQuadrature. Probably not a good
practice to rename classes like this, but this should be much clearer.
